pkg/position_store/cache.go

```diff
@@ -125,19 +123,41 @@ func (cache *defaultPositionCache) Get() (Position, bool, error) {
 	defer cache.positionMutex.Unlock()
 
 	if !cache.exist {
-		position, exist, err := cache.repo.Get(cache.pipelineName)
-		if err != nil && exist {
-			cache.exist = true
+		loaded, err := cache.loadFromRepo()
+		if err != nil {
+			return Position{}, loaded, errors.Trace(err)
 		}
-		return position, cache.exist, errors.Trace(err)
-	}
 
-	if err := cache.position.Validate(); err != nil {
-		return Position{}, true, errors.Trace(err)
+		if loaded {
+			return cache.position, true, nil
+		}
+
+		return Position{}, false, nil
 	}
+
 	return cache.position, true, nil
 }
 
+func (cache *defaultPositionCache) GetWithValueString() (PositionMeta, string, bool, error) {
+	cache.positionMutex.Lock()
+	defer cache.positionMutex.Unlock()
+
+	if !cache.exist {
+		loaded, err := cache.loadFromRepo()
+		if err != nil {
+			return PositionMeta{}, "", loaded, errors.Trace(err)
+		}
+
+		if loaded {
+			return cache.position.PositionMeta, cache.positionValueString, true, nil
+		}
+
+		return PositionMeta{}, "", false, nil
+	}
+
+	return cache.position.PositionMeta, cache.positionValueString, true, nil
```

Comment:
it seems `positionValueString` and `position` are not always synced(that's what we want to avoid). If a `GetWithValueString` called immediately after a `Put` but before `Flush`, it would get inconsistent result. 

How about add `GetEncoder()` to `PositionCacheInterface`? if the caller really want to serialize the position anyway it could do by itself(and in his own goroutine without hold the lock). I suppose this is a rare case.

Reply:
@Ryan-Git changed method name.